src/deepqmc/physics.py

```diff
@@ -46,6 +47,224 @@ def electronic_potential(rs):
+
+    r_en = dists[:, idxs]  # electron-nucleus distances for all electrons and PP nuclei
+
+    # eq. (4) from [Burkatzki et al. 2007]
+    coulomb_term = jnp.einsum(
+        'ij,ki->kji', loc_params[idxs, 0, 1, :], 1 / r_en
+    ) * jnp.exp(jnp.einsum('ij,ki->kji', -loc_params[idxs, 0, 0, :], r_en**2))
+    const_term = jnp.einsum(
+        'ij,kji->kji',
+        loc_params[idxs, 1, 1, :],
+        jnp.exp(jnp.einsum('ij,ki->kji', -loc_params[idxs, 1, 0, :], r_en**2)),
+    )
+    linear_term = jnp.einsum('ij,ki->kji', loc_params[idxs, 2, 1, :], r_en) * jnp.exp(
+        jnp.einsum('ij,ki->kji', -loc_params[idxs, 2, 0, :], r_en**2)
+    )
+
+    # Summation is carried over:
+    # - individual cores in the molecule (idxs dimension)
+    # - individual electrons (1st dimension of 'dists')
+    # - potentially over different coeffs for the term of the same type (ccECP case)
+    pseudopotential = (coulomb_term + const_term + linear_term).sum(axis=(-1, -2, -3))
+
+    return effective_coulomb_potential + pseudopotential
+
+
+@jax.vmap
+def sph2cart(sph, r=1):
+    # This function transforms from spherical to cartesian coordinates.
+    theta = sph[0]
+    phi = sph[1]
+    rsin_theta = r * jnp.sin(theta)
+    x = rsin_theta * jnp.cos(phi)
+    y = rsin_theta * jnp.sin(phi)
+    z = r * jnp.cos(theta)
+    return jnp.array([x, y, z])
+
+
+@jax.jit
+def get_unit_icosahedron_sph():
+    # Basic definition of unit icosahedron vertices in spherical coordinates.
+    unit_icosahedron_sph = []
+    unit_icosahedron_sph.append([0, 0])
+    unit_icosahedron_sph.append([jnp.pi, 0])
+    for j in range(5):
+        unit_icosahedron_sph.append([jnp.arctan(2), jnp.pi / 5 * 2 * j])
+        unit_icosahedron_sph.append([jnp.pi - jnp.arctan(2), jnp.pi / 5 * (2 * j - 1)])
+    return jnp.array(unit_icosahedron_sph)
+
+
+# UNIT_ICOSAHEDRON is (12,3) array of unit icosahedron vertices
+UNIT_ICOSAHEDRON = sph2cart(get_unit_icosahedron_sph())
+QUADRATURE_THETAS = get_unit_icosahedron_sph()[:, 0]
+
+
+def get_quadrature_points(nucleus_position, rs):
+    """
+    Transform electron configuration 'rs' of size (N,3) into an array quadrature points.
+
+    Return array of size (N,12,N,3) that includes all the 12 quadrature point
+    configurations (i.e. reference electron position is shifted to another icosahedron
+    vertex) corresponding to N different reference electron.
+    """
+
+    norm = jnp.linalg.norm(rs - nucleus_position, axis=-1)
+    theta = jnp.arccos((rs - nucleus_position)[..., 2] / norm)
+    phi = jnp.arctan2((rs - nucleus_position)[..., 1], (rs - nucleus_position)[..., 0])
+
+    # rotation matrix about y-axis by angle theta
+    y_rot = jnp.array(
+        [
+            [jnp.cos(theta), jnp.zeros_like(theta), jnp.sin(theta)],
+            [jnp.zeros_like(theta), jnp.ones_like(theta), jnp.zeros_like(theta)],
+            [-jnp.sin(theta), jnp.zeros_like(theta), jnp.cos(theta)],
+        ]
+    )
```

Comment:
Definitions of rotation matrices could be moved to `utils.py`.

Reply:
I agree. Moved in 69ac0f15209eb102c5f0b7dd20a23111ed019aa2